### PR TITLE
Fix bullet check_ghost_overlaps null read

### LINKED ISSUE #21945: Game closes after choose level
**Godot version:**
3.1 latest build

**OS/device including version:**
Archlinux

**Issue description:**
Game closes after choose level in the Godot FPS tutorial project (https://github.com/TwistedTwigleg/Godot_FPS_Tutorial). The main screen open without problems. I also tried with individual level scenes. Seems to be other problem with bullet physics.

handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /usr/lib/libc.so.6(+0x37e00) [0x7fe42766ae00] (??:0)
[2] btCollisionShape::getShapeType() const (??:0)
[3] btCollisionShape::isConvex() const (??:0)
[4] SpaceBullet::check_ghost_overlaps() (??:0)
[5] onBulletTickCallback(btDynamicsWorld*, float) (??:0)
[6] btDiscreteDynamicsWorld::internalSingleStepSimulation(float) (??:0)
[7] btSoftRigidDynamicsWorld::internalSingleStepSimulation(float) (??:0)
[8] btDiscreteDynamicsWorld::stepSimulation(float, int, float) (??:0)
[9] SpaceBullet::step(float) (??:0)
[10] BulletPhysicsServer::step(float) (??:0)
[11] Main::iteration() (??:0)
[12] OS_X11::run() (??:0)
[13] /home/user/git/godot/bin/godot.x11.tools.64(main+0xd5) [0x5604ce77842e] (??:0)
[14] /usr/lib/libc.so.6(__libc_start_main+0xf3) [0x7fe427657223] (??:0)
[15] /home/user/git/godot/bin/godot.x11.tools.64(_start+0x2e) [0x5604ce77828e] (??:0)
-- END OF BACKTRACE --

**Steps to reproduce:**
Open the FPS tutorial on last build and try to play.

### PR DESCRIPTION
**Pull request by muiroc**
fixes #21945

Was reading the issue and in my case otherObject is always non-null and instead get a crash accessing a null other_body_shape or area_shape:

Looks like that when a CollisionShape of a CollisionObject is disabled, that makes the corresponding CollisionObjectBullet::get_bt_shape() return null (this is by design if I understand it correctly)

There is also a ::is_shape_disabled() to check that but i don't know if it ensures that the bt_shape is going to be valid

edit2 about what i didnt get: nvm, reading the file history better, should be because before shapes where always accessed through a compound shape...

Lastly, is it just me or the second and third maps runs with very low fps? But that can't be graphics related since in editor it looks smooth...

### CODE REVIEW

**Review by NewNodeGames — APPROVED** [review]
This solve my issue https://github.com/godotengine/godot/issues/21945

File: modules/bullet/space_bullet.cpp [hunk 1]
```diff
@@ -675,30 +668,43 @@ void SpaceBullet::check_ghost_overlaps() {
...
 
 			bool hasOverlap = false;
 
 			// For each area shape
-			for (y = area->get_shape_count() - 1; 0 <= y; --y) {
-				if (!area->get_bt_shape(y)->isConvex())
+			for (int y = area->get_shape_count() - 1; 0 <= y; --y) {
+				btConvexShape *area_shape = static_cast<btConvexShape *>(area->get_bt_shape(y));
+
+				if (!area_shape) {
```
Comment (AndreaCatania) [inline modules/bullet/space_bullet.cpp]: This check is not correct here, even if it fix the problem. The function above should always return a shape and if it doesn't do that, it means there is a bug somewhere else.

I'll check it this weekend, and thanks you to look on it but feel free to check / Fix it if you want :)

Comment (AndreaCatania) [reply modules/bullet/space_bullet.cpp]: Thanks you too

**Comment by muiroc** [comment]
reporting some findings:

attaching sample debug scene
[bug_ghost_overlaps.zip](https://github.com/godotengine/godot/files/2447701/bug_ghost_overlaps.zip)

get_bt_shape returns null only for CollisionShapes that are disabled since the start, that is, saved as disabled from the editor.
if they initially are enabled, and later disabled from the running scene, then it returns a valid btCollisionShape (the original one) 

but i was wondering why it didnt happen with KinematicBody, so i disabled its CollisionShape and looked for some code using its shape wrappers:
found SpaceBullet::test_body_motion:

inside its first for loop there is this condition:

https://github.com/godotengine/godot/blob/8068d0217a5e74c25f83fe93fa6e077b8d0b3bf5/modules/bullet/space_bullet.cpp#L919-L921

checked with the debugger and the btCollisionShape was null (in p_body->shapes[0].bt_shape)

so without that check there would be a crash!

point is in RigidCollisionObjectBullet::on_shapes_changed the ShapeWrappers create those shapes only when they are active
es on collision_object_bullet.cpp line 349

https://github.com/godotengine/godot/blob/8068d0217a5e74c25f83fe93fa6e077b8d0b3bf5/modules/bullet/collision_object_bullet.cpp#L349-L350
so if shpWrapper->active is not true from the start, bt_shape is going to be null

another issue is that disabling the shapes of the area doesnt trigger the body_exited signal, but this also happens in GodotPhysics

**Comment by muiroc** [comment]
New attempt, maybe I understood what you meant, the null check wasn't right because get_bt_shape would return an allocated pointer even for disabled shapes, so the shapes where going to always be used

Anyway, if more in-depth changes are needed, it's better if I yield since there is a good chance that I'm going to break something ;)

**Comment by AndreaCatania** [comment]
The problem was caused by this line: https://github.com/godotengine/godot/blob/master/modules/bullet/collision_object_bullet.cpp#L355-L357
Where basically I forgot to set `shpWrapper->bt_shape`

However I did some changes and enhanced function order, you can see them here #22759

> another issue is that disabling the shapes of the area doesnt trigger the body_exited signal, but this also happens in GodotPhysics

Is better if you open another issue for this problem.

Thanks you a lot for your interest on fix this :)

**Comment by muiroc** [comment]
@AndreaCatania
Makes sense! Thanks for the explanation, and really appreciate what you are doing, first reimplementing then maintaining the physics must be quite "interesting"!

**Comment by AndreaCatania** [comment]
Thanks you :) it's a pleasure for me

**Pull request closed without merging**